lib/luacrypto_shash.c

```diff
@@ -0,0 +1,343 @@
+	void *state_buf = kmalloc(statesize, gfp);
+	if (!state_buf) return luaL_error(L, "shash:export: failed to allocate state buffer");
+	crypto_shash_export(tfm_ud->kernel_desc, state_buf);
+	lua_pushlstring(
+		L, (const char *)state_buf, statesize
+	);
+	kfree(state_buf);
+	return 1;
+}
+
+/***
+* Imports a previously exported hash state.
+* This overwrites the current hash state and allows resuming a hash calculation.
+* The imported state must be compatible with the current hash algorithm.
+* @function crypto_shash:import
+* @tparam string state The previously exported hash state (binary string).
+* @raise Error on failure or if the provided state length is incorrect for the algorithm.
+*/
+static int luacrypto_shash_tfm_import(lua_State *L) {
+	luacrypto_shash_tfm_t *tfm_ud = luacrypto_check_shash_tfm(L, 1);
+	size_t statelen;
+	const char *state = luaL_checklstring(L, 2, &statelen);
+	unsigned int expected_statesize;
+	int ret;
+	expected_statesize = crypto_shash_statesize(tfm_ud->tfm);
+	luaL_argcheck(
+		L, statelen == expected_statesize,
+		2, "incorrect state length for import"
+	);
+	ret = crypto_shash_import(tfm_ud->kernel_desc, state);
+	if (ret) return luaL_error(L, "shash:import: failed (%d)", ret);
+	return 0;
+}
+
+static const luaL_Reg luacrypto_shash_tfm_mt[] = {
+	{"digestsize", luacrypto_shash_tfm_digestsize},
+	{"setkey", luacrypto_shash_tfm_setkey},
+	{"digest", luacrypto_shash_tfm_digest},
+	{"init", luacrypto_shash_tfm_init},
+	{"update", luacrypto_shash_tfm_update},
+	{"final", luacrypto_shash_tfm_final},
+	{"finup", luacrypto_shash_tfm_finup},
+	{"export", luacrypto_shash_tfm_export},
+	{"import", luacrypto_shash_tfm_import},
+	{"__gc", lunatik_deleteobject},
+	{"__close", lunatik_closeobject},
+	{"__index", lunatik_monitorobject},
+	{NULL, NULL}
+};
+
+/***
+* Lunatik class definition for SHASH objects.
+* This structure binds the C implementation (luacrypto_shash_tfm_t, methods, release function)
+* to the Lua object system managed by Lunatik.
+*/
+static const lunatik_class_t luacrypto_shash_tfm_class = {
+	.name = "crypto_shash_tfm",
+	.methods = luacrypto_shash_tfm_mt,
+	.release = luacrypto_shash_tfm_release,
+	.sleep = true
+};
+
+
+/***
+* Creates a new SHASH object.
+* This is the constructor function for the `crypto_shash` module.
+* @function crypto_shash.new
+* @tparam string algname The name of the hash algorithm (e.g., "sha256", "hmac(sha256)").
+* @treturn crypto_shash The new SHASH object.
+* @raise Error if the TFM object or kernel descriptor cannot be allocated/initialized.
+* @usage
+*   local shash_mod = require("crypto_shash")
+*   local hasher = shash_mod.new("sha256")
+*/
+static int luacrypto_shash_new(lua_State *L) {
+	const char *algname = luaL_checkstring(L, 1);
+	luacrypto_shash_tfm_t *tfm_ud;
+	gfp_t gfp = lunatik_gfp(lunatik_toruntime(L));
+	size_t desc_size;
+	lunatik_object_t *object = lunatik_newobject(
+		L, &luacrypto_shash_tfm_class,
+		sizeof(luacrypto_shash_tfm_t)
+	);
```

Comment:
please, let's keep our code style..
```suggestion
	lunatik_object_t *object = lunatik_newobject(L, &luacrypto_shash_tfm_class,
		sizeof(luacrypto_shash_tfm_t));
```